File: MidAtlantic2015Problems/Hounded/cpp/hounded_sjz.cpp
/*************************************************************
 **                                                          **
 **  Sample solution for "Right-Hand Rule"                   **
 **       Steven J Zeil                                      **
 **       10/29/2005                                         **
 **                                                          **
 *************************************************************/


#include <algorithm>
#include <cassert>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <sstream>
#include <map>

#define NDEBUG 1

using namespace std;

const int MaxMazeSize = 50;

char maze[MaxMazeSize][MaxMazeSize];

struct Location {
	int row;
	int col;

	Location (int r = 0, int c = 0)
	{
		row = r;
		col = c;
	}

	bool operator== (Location loc) const
			{
		return row == loc.row && col == loc.col;
			}
};

Location offsets[4] = {Location(-1, 0), Location(1, 0), Location(0, -1), Location(0,1)};

ostream& operator<< (ostream& out, const Location& loc)
{
	out << '(' << loc.row << ',' << loc.col << ')';
	return out;
}

ostream& operator<< (ostream& out, int dist[MaxMazeSize][MaxMazeSize])
{
	for (int row = 0; row < MaxMazeSize; ++row)
	{
		int printToHere = -1;
		for (int col = 0; col < MaxMazeSize; ++col)
			if (dist[row][col] >= 0)
				printToHere = col;
		if (printToHere >= 0)
		{
			for (int col = 0; col <= printToHere; ++col)
			{
				int d = dist[row][col];
				out << setw(3) << d;
			}
			out << endl;
		}
	}
	return out;
}


list<Location> exits;
Location thiefStart;
Location kennel;


void readMaze (int w, int h, istream& in)
{
	exits.clear();
	string line;
	getline (in, line);
	for (int row = 0; row < h; ++row)
	{
		getline (in, line);
		assert (line.size() <= w);
		for (int col = 0; col < line.size(); ++col) {
			char c = line[col];
			maze[row][col] = c;
			if (c == 'T')
				thiefStart = Location(row, col);
			else if (c == 'K')
				kennel = Location(row, col);
			else if (c == 'E')
				exits.push_back (Location(row, col));
		}
		for (int col = line.size(); col < w; ++col) {
			maze[row][col] = 'X';
		}
	}
}


bool legal(Location loc, int w, int h)
{
	return loc.row >= 0 && loc.row < h
			&& loc.col >= 0 && loc.col < w;
}

void flood (int dist[MaxMazeSize][MaxMazeSize], int w, int h, Location start)
{
	for (int row = 0; row < MaxMazeSize; ++row)
		for (int col = 0; col < MaxMazeSize; ++col)
			dist[row][col] = -1;
	dist[start.row][start.col] = 0;
	queue<Location, list<Location> > q;
	q.push (start);
	while (!q.empty()) {
		Location here = q.front();
		q.pop();
		int d = dist[here.row][here.col];
#ifdef DEBUG
		//cerr << "flood\n" <<  dist;
		//cerr << "here: " << here << "d: " << d << endl;
#endif
		for (Location offset: offsets)
		{
			int dr = offset.row;
			int dc = offset.col;
			Location loc (here.row + dr, here.col + dc);
			if ((dr != 0 || dc != 0) && legal(loc, w, h))
			{
				if (maze[loc.row][loc.col] != 'X' && dist[loc.row][loc.col] < 0)
				{
					dist[loc.row][loc.col] = d + 1;
					q.push(loc);
				}
			}
		}
	}
}


bool floodThief (int w, int h, int kennelDist[MaxMazeSize][MaxMazeSize], int exitDist[MaxMazeSize][MaxMazeSize])
{
#ifdef DEBUG
	cerr << "floodThief\n  kennel\n" <<  kennelDist;
	cerr << "  exit\n"<< exitDist << endl;
#endif
	int dist[MaxMazeSize][MaxMazeSize];
	for (int row = 0; row < MaxMazeSize; ++row)
		for (int col = 0; col < MaxMazeSize; ++col)
			dist[row][col] = -1;
	dist[thiefStart.row][thiefStart.col] = 0;
	queue<Location, list<Location> > q;
	q.push (thiefStart);
	while (!q.empty()) {
		Location here = q.front();
		q.pop();
		int d = dist[here.row][here.col];
		int kd = kennelDist[here.row][here.col];
#ifdef DEBUG
		cerr << "here: " << here << " d: " << d << " kd: " << kd << endl;
#endif

		if (d >= kd)
		{
			// Constable can get here before or at same time as thief
			continue;
		}
		int ed = exitDist[here.row][here.col];
		int thiefCanReachExitOnTurn = d + ed;
		int houndCanReachExitOnTurn = kd + (ed + 1)/2;
		if (thiefCanReachExitOnTurn >= houndCanReachExitOnTurn)
		{
			// Hound can reach the exit before or at same time as thief
			continue;
		}
		if (exitDist[here.row][here.col] == 0)
		{
#ifdef DEBUG
			cerr << "escape at " << here << " on turn " << d << endl;
#endif
			return true;  // Thief escapes
		}

		for (Location offset: offsets)
		{
			int dr = offset.row;
			int dc = offset.col;
			Location loc (here.row + dr, here.col + dc);
			if ((dr != 0 || dc != 0) && legal(loc, w, h))
			{
				if (maze[loc.row][loc.col] != 'X' && dist[loc.row][loc.col] < 0)
				{
					dist[loc.row][loc.col] = d + 1;
					q.push(loc);
				}
			}
		}
#ifdef DEBUG
		cerr << "thief\n" <<  dist;
#endif
	}
	return false;
}


void solve (istream& in)
{
	int w, h;

	int kennelDist[MaxMazeSize][MaxMazeSize];

	while ((in >> w >> h) && (w >= 3) && (h >= 3))
	{
		readMaze (w, h, in);
		bool escaped = false;
		flood (kennelDist, w, h, kennel);
		for (Location exit: exits)
		{
			int exitDist[MaxMazeSize][MaxMazeSize];
			flood (exitDist, w, h, exit);
			escaped = floodThief(w, h, kennelDist, exitDist);
			if (escaped)
				break;
		}
		if (escaped)
		{
			cout << "KEEP IT" << endl;
		}
		else
		{
			cout << "DROP IT" << endl;
		}
	}
}


/*
 * Run the program. Because Eclipse does not support easy redirection of
 * input, if a command line parameter is supplied, treat that as the name
 * of the input file. Otherwise, read from cin.
 */
int main(int argc, char** argv)
{
	if (argc > 1)
	{
		ifstream input (argv[1]);
		solve (input);
	}
	else
	{
		solve (cin);
	}
	return 0;
}

File: MidAtlantic2015Problems/Hounded/java/Hounded_sjz.java
/**
 * Sample solution for the Hounded by Indecision problem.
 * 
 */

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;
import java.util.logging.ConsoleHandler;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.Logger;


public class Hounded_sjz {

	static Logger LOG = Logger.getLogger(Hounded_sjz.class.getName());

	static {
		Handler h = new ConsoleHandler();
		LOG.setLevel(Level.INFO);  // Change to INFO/FINER to toggle debug output
		h.setLevel(Level.FINER);
		LOG.addHandler(h);
	}

	final int MaxMazeSize = 40;

	char[][] maze = new char[MaxMazeSize][MaxMazeSize];

	public class  Location {
		int row;
		int col;

		public Location () {
			row = 0;
			col = 0;
		}

		public Location (int r, int c) {
			row = r;
			col = c;
		}

		public boolean equals (Object obj) {
			Location loc = (Location)obj;
			return row == loc.row && col == loc.col;
		}
		
		public String toString() {
			return "(" + row + "," + col + ")";
		}
	};

	public Location[] offsets = {new Location(-1, 0), new Location(1, 0),
			new Location(0, -1), new Location(0,1)};


	private int w;
	private int h;
	private Scanner input;
	
	public Hounded_sjz(int w0, int h0, Scanner input0) {
		w = w0;
		h = h0;
		input = input0;
	}


	public String toString(int[][] dist) {
		StringBuffer out = new StringBuffer();
		for (int row = 0; row < MaxMazeSize; ++row) {
			int printToHere = -1;
			for (int col = 0; col < MaxMazeSize; ++col) {
				if (dist[row][col] >= 0) {
					printToHere = col;
				}
			}
			if (printToHere >= 0)
			{
				for (int col = 0; col <= printToHere; ++col)
				{
					int d = dist[row][col];
					String ds = new Integer(d).toString();
					for (int i = ds.length(); i < 3; ++i)
						out.append(' ');
					out.append(ds);
				}
				out.append ("\n");
			}
		}
		return out.toString();
	}


	List<Location> exits;
	Location thiefStart;
	Location kennel;


	void readMaze (int w, int h, Scanner in)
	{
		exits = new ArrayList<Location>();
		String line = in.nextLine();
		for (int row = 0; row < h; ++row)
		{
			line = in.nextLine();
			assert (line.length() <= w);
			for (int col = 0; col < line.length(); ++col) {
				char c = line.charAt(col);
				maze[row][col] = c;
				if (c == 'T')
					thiefStart = new Location(row, col);
				else if (c == 'K')
					kennel = new Location(row, col);
				else if (c == 'E')
					exits.add (new Location(row, col));
			}
			for (int col = line.length(); col < w; ++col) {
				maze[row][col] = 'X';
			}
		}
	}


	boolean legal(Location loc, int w, int h)
	{
		return loc.row >= 0 && loc.row < h
				&& loc.col >= 0 && loc.col < w;
	}

	void flood (int[][] dist, int w, int h, Location start)
	{
		for (int row = 0; row < MaxMazeSize; ++row)
			for (int col = 0; col < MaxMazeSize; ++col)
				dist[row][col] = -1;
		dist[start.row][start.col] = 0;
		Queue<Location> q = new LinkedList<Location>();
		q.add (start);
		while (!q.isEmpty()) {
			Location here = q.element();
			q.remove();
			int d = dist[here.row][here.col];
			if (LOG.isLoggable(Level.FINE)) {
				LOG.log(Level.FINE, "in flood, dist=" + dist);
				LOG.log(Level.FINE, toString(dist));
			}
			
			for (Location offset: offsets)
			{
				int dr = offset.row;
				int dc = offset.col;
				Location loc = new Location(here.row + dr, here.col + dc);
				if ((dr != 0 || dc != 0) && legal(loc, w, h))
				{
					if (maze[loc.row][loc.col] != 'X' && dist[loc.row][loc.col] < 0)
					{
						dist[loc.row][loc.col] = d + 1;
						q.add(loc);
					}
				}
			}
		}
	}


	boolean floodThief (int w, int h, int[][] kennelDist, int[][] exitDist)
	{
		if (LOG.isLoggable(Level.FINE)) {
			LOG.log(Level.FINE, "in floodThief, kennelDist:");
			LOG.log(Level.FINE, toString(kennelDist));
			LOG.log(Level.FINE, "in floodThief, exitDist:");
			LOG.log(Level.FINE, toString(exitDist));
		}
		int[][] dist = new int[MaxMazeSize][MaxMazeSize];
		for (int row = 0; row < MaxMazeSize; ++row)
			for (int col = 0; col < MaxMazeSize; ++col)
				dist[row][col] = -1;
		dist[thiefStart.row][thiefStart.col] = 0;
		Queue<Location> q = new LinkedList<Location>();
		q.add (thiefStart);
		while (!q.isEmpty()) {
			Location here = q.element();
			q.remove();
			int d = dist[here.row][here.col];
			int kd = kennelDist[here.row][here.col];
			LOG.log(Level.FINE,  "here: " + here + " d: " + d + " kd: " + kd);

			if (d >= kd) {
				// Constable can get here before or at same time as thief
				continue;
			}
			int ed = exitDist[here.row][here.col];
			int thiefCanReachExitOnTurn = d + ed;
			int houndCanReachExitOnTurn = kd + (ed + 1)/2;
			if (thiefCanReachExitOnTurn >= houndCanReachExitOnTurn) {
				// Hound can reach the exit before or at same time as thief
				continue;
			}
			if (exitDist[here.row][here.col] == 0) {
				LOG.log(Level.FINE,  "escape at " + here + " on turn " + d);
				return true;  // Thief escapes
			}

			for (Location offset: offsets)
			{
				int dr = offset.row;
				int dc = offset.col;
				Location loc = new Location (here.row + dr, here.col + dc);
				if ((dr != 0 || dc != 0) && legal(loc, w, h))
				{
					if (maze[loc.row][loc.col] != 'X' && dist[loc.row][loc.col] < 0)
					{
						dist[loc.row][loc.col] = d + 1;
						q.add(loc);
					}
				}
			}
			if (LOG.isLoggable(Level.FINE)) {
				LOG.log(Level.FINE,  "thief");
				LOG.log(Level.FINE, toString(dist));
			}
		}
		return false;
	}


	void solve ()
	{
		int[][] kennelDist = new int[MaxMazeSize][MaxMazeSize];

		readMaze (w, h, input);
		boolean escaped = false;
		flood (kennelDist, w, h, kennel);
		for (Location exit: exits)
		{
			int exitDist[][] = new int[MaxMazeSize][MaxMazeSize];
			flood (exitDist, w, h, exit);
			escaped = floodThief(w, h, kennelDist, exitDist);
			if (escaped)
				break;
		}
		if (escaped)
		{
			System.out.println("KEEP IT");
		}
		else
		{
			System.out.println("DROP IT");
		}
	}





	/**
	 * Run the program.  Because redirecting input is impossible in Eclipse,
	 *    if a command line parameter is given, it is treated as the name
	 *    of an input file. If no command parameter is given, input is taken
	 *    from System.in. 
	 * @param args  Command line parameters  
	 * @throws IOException 
	 */
	public static void main (String[] args) throws IOException
	{
		Scanner input;
		if (args.length == 0)
			input = new Scanner(System.in);
		else
			input = new Scanner(new File(args[0]));

		int W = input.nextInt();
		while (W >= 3) {
			int H = input.nextInt();
			Hounded_sjz problem = new Hounded_sjz (W, H, input);
			problem.solve();
			W = input.nextInt();
		}
		input.close();
	}
}
